static/app/views/starfish/types.tsx

```diff
@@ -4,6 +4,11 @@ import {DiscoverDatasets} from 'sentry/utils/discover/types';
 import type {FieldDefinition} from 'sentry/utils/fields';
 import {FieldKind, FieldValueType} from 'sentry/utils/fields';
 
+export interface ErrorLike {
```

Comment:
Instead of using `ErrorLike`, could we type this a little stronger?

From what I can see the instances of this seem to come from `useGenericDiscoverQuery`, which puts an error into `QueryError`, which .

https://github.com/getsentry/sentry/blob/bb76da49db424f24f6caa9e1dbc206fc69a1f312/static/app/utils/discover/genericDiscoverQuery.tsx#L436

https://github.com/getsentry/sentry/blob/bb76da49db424f24f6caa9e1dbc206fc69a1f312/static/app/utils/discover/genericDiscoverQuery.tsx#L25

I would try to see if we can use `QueryError` here and then update `QueryError` to extend the `Error` builtin so we get stacktraces and can actually pass the error into sentry. (`QueryError` should also prob use `error.cause`).

If we don't elect to update `QueryError`, another possible approach here is for us to have a starfish module specific error that we link the `errorLike` obj returned from the queries via [`error.cause`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error/cause).

So something like:

```ts
// inside useSpanMetricsSeries

const result = useWrappedDiscoverTimeseriesQuery ...

// naming tdb - feel free to pick what feels right to you
const starfishError = new StarfishError(result.error)

// starfishError is then passed down here - 
```

not totally happy with above, but hopefully gives you the broad strokes.

Reply:
Great question, I was torn on this. I want to be permissive in what I accept here (e.g., `QueryError` is a discover-only concept and I want to accept all kinds of errors) so I went with a weaker type on purpose, rather than tying myself to `QueryError`. The other thing I was hoping is to use `Error` because even though it's more restrictive, it's at least a _common_ type, but `QueryError` doesn't intersect with it nicely.

Maybe `ErrorLike` is too weak? I like the idea of good coverage of error-like properties like `cause` but I have the sense that the more strict I am, the more annoying it'll be to set these errors up in unit tests, and the less flexible it'll be (e.g., people won't be able to easily pass `error={` in case of some custom code-land error when processing data)

What do you think?

Reply:
Ideally if `QueryError` just extends `Error`, we can have all of the components take the `Error` type as the prop. This should also mean it's pretty easy to unit test, given people can just pass in `new Error()` to test. What makes `QueryError` not intersect well with it right now?

```
// maybe even rename to DiscoverQueryError?
export class QueryError extends Error {
  constructor(message, originalError?: any) {
    super(message);
    this.name = "QueryError";
    this.cause = originalError;
  }
}
```

The best part of using `Error` built-in is that we get stacktraces with them, so if these errors are thrown to Sentry or even just logged to console they are a lot more useful in terms of debugging context.

Reply:
@AbhiPrasad ah, I like `extend`ing `Error`! I was wary of trying to make the `QueryError` interface completely match `Error` which is what I meant by "doesn't intersect nicely". Extending `Error` works great, even added a cheeky spec to prove that point 👍🏻 No more `ErrorLike`